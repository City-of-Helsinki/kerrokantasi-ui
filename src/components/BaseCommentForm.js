/* eslint-disable react/no-did-mount-set-state */
import React from 'react';
import PropTypes from 'prop-types';
import {injectIntl, intlShape, FormattedMessage, } from 'react-intl';
import { Button, Checkbox, FormControl, FormGroup, ControlLabel, Alert } from 'react-bootstrap';
import classnames from 'classnames';
import {connect} from 'react-redux';
import uuid from 'uuid/v1';
import Icon from '../utils/Icon';
import {getImageAsBase64Promise} from '../utils/hearing';
import CommentDisclaimer from './CommentDisclaimer';
import {get, includes} from 'lodash';
import QuestionResults from './QuestionResults';
import QuestionForm from './QuestionForm';
import {localizedNotifyError} from "../utils/notify";
import {getSectionCommentingErrorMessage, isSectionCommentingMapEnabled} from "../utils/section";
import {Polygon, GeoJSON, Polyline, Circle} from 'react-leaflet';
// eslint-disable-next-line import/no-unresolved
import urls from '@city-assets/urls.json';
// eslint-disable-next-line import/no-unresolved
import localization from '@city-i18n/localization.json';
import leafletMarkerIconUrl from '../../assets/images/leaflet/marker-icon.png';
import {getCorrectContrastMapTileUrl} from "../utils/map";
import Leaflet, {LatLng} from 'leaflet';
import leafletMarkerShadowUrl from "../../assets/images/leaflet/marker-shadow.png";
import leafletMarkerRetinaIconUrl from "../../assets/images/leaflet/marker-icon-2x.png";
import CommentFormMap from "./CommentFormMap/CommentFormMap";

Leaflet.Marker.prototype.options.icon = new Leaflet.Icon({
  iconUrl: leafletMarkerIconUrl,
  shadowUrl: leafletMarkerShadowUrl,
  iconRetinaUrl: leafletMarkerRetinaIconUrl,
  iconSize: [25, 41],
  iconAnchor: [13, 41],
});
export class BaseCommentForm extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      collapsed: true,
      commentText: "",
      nickname: props.defaultNickname || '',
      imageTooBig: false,
      images: [],
      pinned: false,
      showAlert: true,
      hideName: false,
      geojson: {},
      mapCommentText: "",
    };
    this.getSelectedImagesAsArray = this.getSelectedImagesAsArray.bind(this);
  }

  componentDidMount = () => {
    if (this.isUserAdmin()) {
      this.setState({ nickname: this.props.user.displayName });
    }
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.collapseForm && nextProps.collapseForm) {
      this.clearCommentText();
      this.toggle();
    }
    if (this.props.defaultNickname === '' && nextProps.defaultNickname !== '' && !this.isUserAdmin()) {
      this.setState({nickname: nextProps.defaultNickname});
    }
    if (this.isUserAdmin() && nextProps.user && nextProps.user.displayName) {
      this.setState({ nickname: nextProps.user.displayName });
    }
  }

  toggle() {
    const {canComment, section} = this.props;
    if (canComment) {
      this.setState({
        collapsed: !this.state.collapsed,
        commentText: "",
        nickname: this.props.defaultNickname || '',
        imageTooBig: false,
        images: [],
        pinned: false,
        showAlert: true,
        hideName: false,
        mapCommentText: "",
      });
      if (this.props.onOverrideCollapse instanceof Function) {
        this.props.onOverrideCollapse();
      }
    } else {
      localizedNotifyError(getSectionCommentingErrorMessage(section));
    }
  }

  handleTextChange(event) {
    this.setState({commentText: event.target.value});
  }

  handleNicknameChange(event) {
    this.setState({nickname: event.target.value});
  }

  clearCommentText() {
    this.setState({commentText: ""});
  }

  submitComment() {
    const pluginComment = this.getPluginComment();
    let pluginData = this.getPluginData();
    let nickname = (this.state.nickname === "" ? this.props.nicknamePlaceholder : this.state.nickname);
    let commentText = (this.state.commentText === null ? '' : this.state.commentText);
    let geojson = this.state.geojson;
    let label = null;
    let images = this.state.images;
    let pinned = this.state.pinned;
    let mapCommentText = this.state.mapCommentText;

    // plugin comment will override comment fields, if provided
    if (pluginComment) {
      commentText = pluginComment.content || commentText;
      nickname = pluginComment.author_name || nickname;
      pluginData = pluginComment.plugin_data || pluginData;
      label = pluginComment.label || null;
      images = pluginComment.image ? [pluginComment.image] : images;
      geojson = pluginComment.geojson || geojson;
      pinned = pluginComment.pinned || null;
      mapCommentText = pluginComment.mapCommentText || mapCommentText;
    } else if (pluginData && typeof pluginData !== "string") {
      // this is for old-fashioned plugins with only data
      pluginData = JSON.stringify(pluginData);
    }
    this.props.onPostComment(
      commentText,
      nickname,
      pluginData,
      geojson,
      label,
      images,
      pinned,
      mapCommentText,
    );
    this.setState({
      collapsed: false,
      commentText: "",
      nickname: this.props.defaultNickname || '',
      imageTooBig: false,
      images: [],
      pinned: false,
      showAlert: true,
      hideName: false,
      geojson: {},
      mapCommentText: "",
    });
  }

  handleChange(event) {
    const imagePromisesArray = [];
    const images = [];
    this.isImageTooBig(event.target.files);

    for (let _i = 0; _i < this.refs.images.files.length; _i += 1) {
      imagePromisesArray.push(getImageAsBase64Promise(this.refs.images.files[_i]));
    }

    Promise.all(imagePromisesArray).then((arrayOfResults) => {
      for (let _i = 0; _i < this.refs.images.files.length; _i += 1) {
        const imageObject = {title: "Title", caption: "Caption"};

        imageObject.image = arrayOfResults[_i];
        images.push(imageObject);
      }

      this.setState({images});
    });
  }

  /**
   * Determines whether the logged in user is admin or not.
   * The array in users with key adminOrganizations should be of length > 0
   */
  isUserAdmin = () => (
    this.props.user
    && Array.isArray(this.props.user.adminOrganizations)
    && this.props.user.adminOrganizations.length > 0
  );

  getPluginData() {  // eslint-disable-line class-methods-use-this
    return undefined;
  }

  getPluginComment() {  // eslint-disable-line class-methods-use-this
    return undefined;
  }

  getSelectedImagesAsArray(files) { // eslint-disable-line class-methods-use-this
    const imagesArray = [];
    for (let _i = 0; _i < files.length; _i += 1) {
      imagesArray.push(files[_i]);
    }
    return imagesArray;
  }

  /**
   * When user type is admin, an alert is shown, use this method to close the alert.
   */
  handleCloseAlert = () => {
    this.setState((prevState) => ({
      showAlert: !prevState.showAlert,
    }));
  }

  /**
   * When logged in as admin, user may chose to hide their identity.
   */
  handleToggleHideName = () => {
    this.setState((prevState) => ({
      nickname: !prevState.hideName ? this.props.intl.formatMessage({ id: 'employee' }) : this.props.user.displayName,
      hideName: !prevState.hideName,
    }));
  }

  /**
   * Toggle the pinning of comment
   */
  handleTogglePin = () => {
    this.setState((prevState) => ({
      pinned: !prevState.pinned,
    }));
  }

  isImageTooBig(images) { // eslint-disable-line class-methods-use-this
    let isImageTooBig = false;
    Array.from(images).forEach((image) => {
      if (image.size > 1000000) {
        isImageTooBig = true;
      }
    });
    if (isImageTooBig) {
      this.setState({imageTooBig: true});
    } else {
      this.setState({imageTooBig: false});
    }
  }

  /**
   * When admin user is posting a comment, we will show a closeable warning.
   */
  renderAdminWarning = () => (
    <Alert bsStyle="warning">
      <div className="comment-form__comment-alert">
        <div className="comment-form__comment-alert__alert-icon">
          <Icon name="info-circle" size="lg" />
        </div>
        <span className="comment-form__comment-alert__alert-message"><FormattedMessage id="adminCommentMessage"/></span>
        <div className="comment-form__comment-alert__alert-close">
          <Icon name="close" onClick={this.handleCloseAlert}/>
        </div>
      </div>
    </Alert>
  );

  /**
   * Render the checkbox to hide user name and identitiy for admin user.
   */
  renderHideNameOption = () => (
    <Checkbox checked={this.state.hideName} key={uuid()} onChange={this.handleToggleHideName}>
      <FormattedMessage id="hideName"/>
    </Checkbox>
  );

  /**
   * For admins, there is slightly different form.
   */
  renderFormForAdmin =() => {
    const { user } = this.props;
    const organization = this.isUserAdmin() && user.adminOrganizations[0];
    return (
      <FormGroup>
        <FormControl
          type="text"
          placeholder={this.props.nicknamePlaceholder}
          value={this.state.nickname}
          onChange={this.handleNicknameChange.bind(this)}
          maxLength={32}
          disabled
        />
        <FormControl
          type="text"
          placeholder={this.props.intl.formatMessage({ id: 'organization' })}
          value={organization || ''}
          onChange={() => {}}
          maxLength={32}
          disabled
        />
      </FormGroup>
    );
  }

  /**
   * If an admin type of user is posting comment, the form is slightly different.
   * @returns {JSX<Component>}
   */
  renderNameFormForUser = () => {
    const isAdminUser = this.isUserAdmin();
    const headingId = isAdminUser ? 'nameAndOrganization' : 'nickname';

    return (
      <React.Fragment>
        <label htmlFor="commentNickname" className="h4">
          <FormattedMessage id={headingId} />
        </label>
        { isAdminUser && this.state.showAlert && this.renderAdminWarning() }
        { isAdminUser && this.renderHideNameOption() }
        {
          isAdminUser
          ? (
            <div className="comment-form__group-admin">
              { this.renderFormForAdmin() }
            </div>
          )
          : (
            <FormGroup>
              <FormControl
                id="commentNickname"
                type="text"
                placeholder={this.props.nicknamePlaceholder}
                value={this.state.nickname}
                onChange={this.handleNicknameChange.bind(this)}
                maxLength={32}
              />
            </FormGroup>
          )
        }
      </React.Fragment>
    );
  }

  /**
   * When user is of admin type, they are allowed to pin a comment to top.
   * In the form, an icon can be shown to pin or unpin the comment.
   */
  renderPinUnpinIcon = () => (
    <Button
      className={classnames([
        'comment-form__heading-container__pin__icon',
        {
        'comment-form__heading-container__pin__pin-comment': !this.state.pinned,
        'comment-form__heading-container__pin__unpin-comment': this.state.pinned
        }
      ])}
      onClick={this.handleTogglePin}
    />
  );
  onDrawCreate = (event) => {
    this.setState({geojson: event.layer.toGeoJSON().geometry});
  }

  onDrawDelete = () => {
    this.setState({geojson: null});
  }

  handleMapTextChange(event) {
    this.setState({mapCommentText: event.target.value});
  }
  getMapBorder() {
    const {hearingGeojson} = this.props;

    if (hearingGeojson && hearingGeojson.type !== 'Point') {
      const contents = [];
      if (hearingGeojson.type === 'FeatureCollection') {
        hearingGeojson.features.forEach((feature) => {
          contents.push(this.getMapElement(feature));
        });
      } else {
        contents.push(this.getMapElement(hearingGeojson));
      }
      return contents;
    }
    return null;
  }
  getMapElement(geojson) {
    switch (geojson.type) {
      case "Polygon": {
        // XXX: This only supports the _first_ ring of coordinates in a Polygon
        const latLngs = geojson.coordinates[0].map(([lng, lat]) => new LatLng(lat, lng));
        return (<Polygon key={Math.random()} positions={latLngs} color="transparent" />);
      }
      case "Point": {
        const latLngs = new LatLng(geojson.coordinates[1], geojson.coordinates[0]);
        return (
          <Circle center={latLngs} radius={10} color="transparent" />
        );
      }
      case "LineString": {
        const latLngs = geojson.coordinates.map(([lng, lat]) => new LatLng(lat, lng));
        return (<Polyline key={Math.random()} positions={latLngs} color="transparent" />);
      }
      case "Feature": {
        return (this.getMapElement(geojson.geometry));
      }
      default:
        // This should never happen
        return (<GeoJSON data={geojson} key={JSON.stringify(geojson)} color="transparent" />);
    }
  }

  getMapCenter() {
    const {hearingGeojson} = this.props;
    let center;
    if (hearingGeojson && hearingGeojson.type === 'Point') {
      center = new LatLng(hearingGeojson.coordinates[1], hearingGeojson.coordinates[0]);
    } else {
      center = new LatLng(localization.mapPosition[0], localization.mapPosition[1]);
    }
    return center;
  }

  getMapContrastTiles() {
    const {isHighContrast, language} = this.props;
    return getCorrectContrastMapTileUrl(urls.rasterMapTiles,
      urls.highContrastRasterMapTiles, isHighContrast, language);
  }

  render() {
    const {language, section, onChangeAnswers, answers, loggedIn, closed, user, isReply, canComment} = this.props;
    if (!this.props.overrideCollapse && this.state.collapsed) {
      return (
        <Button onClick={this.toggle.bind(this)} bsStyle="primary" bsSize="large" block>
          <Icon name="comment"/> <FormattedMessage id="addComment"/>
        </Button>
      );
    }
    return (
      <div className="comment-form">
        <form>
          <h2><FormattedMessage id="writeComment"/></h2>
          {
            !isReply &&
            section.questions.map((question) => {
              const canShowQuestionResult =
                closed || (loggedIn && includes(get(user, "answered_questions"), question.id));
              return canShowQuestionResult
                ? <QuestionResults key={question.id} question={question} lang={language} />
                : null;
            })
          }
          {
            !isReply &&
            section.questions.map((question) => {
              const canShowQuestionForm = !closed && !includes(get(user, "answered_questions"), question.id);
              return canShowQuestionForm
                ? (
                  <QuestionForm
                    key={question.id}
                    canAnswer={canComment}
                    answers={answers.find(answer => answer.question === question.id)}
                    onChange={onChangeAnswers}
                    question={question}
                    lang={language}
                  />
                )
                : null;
            })
          }
          <div className="comment-form__heading-container">
            <div className="comment-form__heading-container__title">
              <label htmlFor="commentTextField" className="h4">
                <FormattedMessage id="writeComment"/>
              </label>
            </div>
            {
              this.isUserAdmin()
              && !isReply
              && (
                <div className="comment-form__heading-container__pin">
                  { this.renderPinUnpinIcon() }
                </div>
              )
            }
          </div>
          <FormControl
            autoFocus
            componentClass="textarea"
            value={this.state.commentText}
            onChange={this.handleTextChange.bind(this)}
            id="commentTextField"
          />
          {isSectionCommentingMapEnabled(user, section) && (
            <div className="comment-form__map-container" style={{ marginTop: 20}}>
              <div>
                <label htmlFor="commentMapAddress">
                  <FormattedMessage id="commentMapTitle" />
                </label>
              </div>
              <FormattedMessage id="commentMapInstructions">
                {instr => <span style={{fontSize: 13}}>{instr}</span>}
              </FormattedMessage>
              <div className="map-padding">
                <CommentFormMap
                  center={this.getMapCenter()}
                  mapBounds={localization.mapBounds || null}
                  mapTileUrl={this.getMapContrastTiles()}
                  onDrawCreate={this.onDrawCreate}
                  onDrawDelete={this.onDrawDelete}
                  contents={this.getMapBorder()}
                  tools={section.commenting_map_tools}
                  language={language}
                />
              </div>
              <FormGroup>
                <ControlLabel htmlFor="map_text">
                  <FormattedMessage id="commentMapAdditionalInfo"/>
                </ControlLabel>
                <FormControl
                  id="map_text"
                  type="text"
                  value={this.state.mapCommentText}
                  onChange={this.handleMapTextChange.bind(this)}
                  maxLength={128}
                />
              </FormGroup>
            </div>
          )}

          <div className="comment-form__selected-images">
            {this.state.imageTooBig
              ? (
                <div className="comment-form__image-too-big">
                  <FormattedMessage id="imageSizeError"/>
                </div>
              )
              : this.state.images.map(
                (image, key) =>
                  <img
                    style={{ marginRight: 10 }}
                    alt=""
                    src={image.image}
                    width={image.width < 100 ? image.width : 100}
                    height={image.height < 100 ? image.width : 100}
                    key={key + Math.random()} //eslint-disable-line
                  />)
            }
          </div>
          <FormGroup className="comment-form__file">
            <ControlLabel><FormattedMessage id="add_images"/></ControlLabel>
            <div className="comment-form__select-file">
              <input
                type="file"
                ref="images"
                id="fileInput"
                multiple
                className="custom-file-input"
                onChange={(event) => this.handleChange(event)}
              />
              <label className="btn btn-default btn-sm" htmlFor="fileInput">
                <FormattedMessage id="choose_images" />
              </label>
            </div>
            <span style={{fontSize: 13, marginTop: 20}}><FormattedMessage id="multipleImages"/></span>
          </FormGroup>
          { this.renderNameFormForUser() }
          <div className="comment-buttons clearfix">
            <Button
              bsStyle="default"
              onClick={this.toggle.bind(this)}
            >
              <FormattedMessage id="cancel"/>
            </Button>
            <Button
              bsStyle="primary"
              disabled={!this.state.commentText || this.state.imageTooBig}
              onClick={this.submitComment.bind(this)}
            >
              <FormattedMessage id="submit"/>
            </Button>
          </div>
          <CommentDisclaimer/>
        </form>
      </div>
    );
  }
}

BaseCommentForm.propTypes = {
  canComment: PropTypes.bool,
  onPostComment: PropTypes.func,
  onOverrideCollapse: PropTypes.func,
  intl: intlShape.isRequired,
  collapseForm: PropTypes.bool,
  defaultNickname: PropTypes.string,
  overrideCollapse: PropTypes.bool,
  nicknamePlaceholder: PropTypes.string,
  section: PropTypes.object,
  language: PropTypes.string,
  onChangeAnswers: PropTypes.func,
  answers: PropTypes.array,
  loggedIn: PropTypes.bool,
  closed: PropTypes.bool,
  user: PropTypes.object,
  isReply: PropTypes.bool,
  isHighContrast: PropTypes.bool,
  hearingGeojson: PropTypes.object
};

BaseCommentForm.defaultProps = {
  defaultNickname: '',
  overrideCollapse: false,
  onOverrideCollapse: () => {},
  isReply: false,
};

const mapStateToProps = (state) => ({
  isHighContrast: state.accessibility.isHighContrast,
});
const WrappedBaseCommentForm = connect(mapStateToProps, null)(injectIntl(BaseCommentForm));
export default WrappedBaseCommentForm;
