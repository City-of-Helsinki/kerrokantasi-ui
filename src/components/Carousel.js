import React from 'react';
import PropTypes from 'prop-types';
import Slider from 'react-slick';
import getAttr from '../utils/getAttr';
import isEmpty from 'lodash/isEmpty';
import findIndex from 'lodash/findIndex';
import {getSectionURL} from '../utils/section';
import {withRouter} from 'react-router-dom';
import { HashLink as Link } from 'react-router-hash-link';
import OverviewMap from './OverviewMap';

export class SectionCarousel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      slideCount: this.getSlideCount()
    };
  }

  componentDidMount() {
    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  handleResize = () => {
    this.setState({slideCount: this.getSlideCount()});
  }

  getSlideCount = () => {
    if (typeof window !== 'undefined' && window.innerWidth < 400) {
      return 1;
    }

    if (typeof window !== 'undefined' && window.innerWidth < 500) {
      return 2;
    }

    if (typeof window !== 'undefined' && window.innerWidth < 768) {
      return 3;
    }

    return 4;
  }

  render() {
    const {hearing, match: {params}, language} = this.props;
    const {slideCount} = this.state;
    const sectionsWithoutClosure = hearing.sections.filter((section) => section.type !== 'closure-info');

    return (
      <div className="carousel-container">
        {hearing.geojson && <HearingMap hearing={hearing} />}
        <div className="slider-container">
          <Slider
            className="slider"
            ref={slider => {
              this.slider = slider;
            }}
            infinite={false}
            focusOnSelect
            slidesToShow={slideCount}
            autoplay={false}
            initialSlide={params.sectionId ? findIndex(sectionsWithoutClosure, (section) => section.id === params.sectionId) : 0}
          >
            {sectionsWithoutClosure.map(
              (section) => <div key={section.id}><SliderItem hearingTitle={hearing.title} url={section.type === 'main' ? `/${hearing.slug}#start` : getSectionURL(hearing.slug, section) + '#start'} language={language} section={section} /></div>)}
          </Slider>
        </div>
      </div>
    );
  }
}

SectionCarousel.propTypes = {
  hearing: PropTypes.object,
  match: PropTypes.object,
  language: PropTypes.string
};

export default withRouter(SectionCarousel);

const HearingMap = ({hearing}) => {
  return (
    <div className="carousel-map-container">
      <div className="carousel-map">
        <OverviewMap hearings={[hearing]} style={{width: '100%', height: '200px'}} hideIfEmpty />
      </div>
    </div>
  );
};

HearingMap.propTypes = {
  hearing: PropTypes.object
};

const SliderItem = ({section, url, language, hearingTitle}) => {
  return (
    <div className="slider-item">
      <Link to={url}>
        {!isEmpty(section.images) && <img className="slider-image" src={section.images[0].url} alt="Section"/>}
        <div className="slider-item-content">
          <div className="slider-item-title">{section.type === 'main' ? getAttr(hearingTitle, language) : getAttr(section.title, language)}</div>
        </div>
      </Link>
    </div>
  );
};


SliderItem.propTypes = {
  section: PropTypes.object,
  url: PropTypes.string,
  language: PropTypes.string,
  hearingTitle: PropTypes.object
};
