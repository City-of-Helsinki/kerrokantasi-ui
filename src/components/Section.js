/* eslint-disable jsx-a11y/no-static-element-interactions */
import React from 'react';
import SortableCommentList from './SortableCommentList';
import {Link} from 'react-router';
import Icon from '../utils/Icon';
import {isSpecialSectionType, userCanComment} from '../utils/section';
import classNames from 'classnames';
import MapdonHKRPlugin from './plugins/legacy/mapdon-hkr';
import MapdonKSVPlugin from './plugins/legacy/mapdon-ksv';
import MapQuestionnaire from './plugins/MapQuestionnaire';
import Alert from 'react-bootstrap/lib/Alert';
import getAttr from '../utils/getAttr';

function getImageList(section, language) {
  if (section.type === "main") { // Main section images aren't rendered here atleast atm.
    return null;
  }
  return section.images.map((image) => (<div key={image.url}>
    <img
      className="img-responsive"
      alt={getAttr(image.title, language)}
      title={getAttr(image.title, language)}
      src={image.url}
    />
    <div className="image-caption">{getAttr(image.caption, language)}</div>
  </div>));
}


export default class Section extends React.Component {
  constructor(props) {
    super(props);
    this.state = {collapsed: true};
  }

  // componentDidMount() {
  //   // if (!this.isCollapsible()) {  // Trigger immediate comment load for uncollapsible sections
  //   //   this.loadComments();
  //   // }
  // }

  componentWillUpdate(nextProps, nextState) {
    if (this.state.collapsed && !nextState.collapsed) {  // (Re-)load comments when uncollapsing
      // this.loadComments();
    }
  }

  onPostComment(text, authorName, pluginData, geojson, label, images) {
    const {section} = this.props;
    const commentData = {text, authorName, pluginData, geojson, label, images};
    if (this.props.onPostComment) {
      this.props.onPostComment(section.id, commentData);
    }
  }

  onPostVote(commentId) {
    const {section} = this.props;
    this.props.onPostVote(commentId, section.id);
  }

  toggle() {
    this.setState({collapsed: !this.state.collapsed});
  }

  loadComments() {
    const {section} = this.props;
    if (this.props.loadSectionComments) {
      this.props.loadSectionComments(section.id);
    }
  }

  getTitleDiv(collapsed, collapsible) {
    const {section, linkTo} = this.props;
    const {language} = this.context;
    if (section.type === "main") {
      return null;
    }
    if (section.type === "closure-info") {
      return (
        <h3 className="section-title">
          {getAttr(section.title, language)}
        </h3>
      );
    }
    const iconName = (collapsed ? "chevron-right" : "chevron-down");

    if (linkTo) {
      return (
        <h2 className="section-title" onClick={this.toggle.bind(this)}>
          <Link to={linkTo}>
            {collapsible ? (<span><Icon name={iconName}/>&nbsp;</span>) : null}
            {getAttr(this.props.section.title, language)}
          </Link>
          {collapsed ? (
            <div className="section-title-comments">
              <Icon name="comment-o"/>&nbsp;{section.n_comments}
            </div>
          ) : null}
        </h2>
      );
    }

    return (
      <h2 className="section-title" onClick={this.toggle.bind(this)}>
        {collapsible ? (<span><Icon name={iconName}/>&nbsp;</span>) : null}
        {getAttr(this.props.section.title, language)}
        {collapsed ? (
          <div className="section-title-comments">
            <Icon name="comment-o"/>&nbsp;{section.n_comments}
          </div>
        ) : null}
      </h2>
    );
  }

  renderPluginContent(section) {
    if (!this.props.showPlugin) {
      return null;
    }
    const {user} = this.props;
    const comments = this.props.comments ? this.props.comments.data : [];
    if (typeof window === 'undefined' || !section.plugin_identifier) {
      return null;
    }
    switch (section.plugin_identifier) {
      case "mapdon-hkr":
        return (
          <MapdonHKRPlugin
            data={section.plugin_data}
            onPostComment={this.onPostComment.bind(this)}
          />
        );
      case "mapdon-ksv":
        return (
          <MapdonKSVPlugin
            data={section.plugin_data}
            onPostComment={this.onPostComment.bind(this)}
            pluginPurpose="postComments"
            canSetNickname={user === null}
          />
        );
      case "map-questionnaire":
        return (
          <MapQuestionnaire
            data={section.plugin_data}
            onPostComment={this.onPostComment.bind(this)}
            onPostVote={this.onPostVote.bind(this)}
            comments={comments}
            pluginPurpose="postComments"
            canSetNickname={user === null}
            displayCommentBox={false}
            pluginSource={section.plugin_iframe_url}
          />
        );
      default:
        return <Alert>I do not know how to render the plugin {section.plugin_identifier}</Alert>;
    }
  }

  isCollapsible() {
    const {section, isCollapsible} = this.props;
    const hasPlugin = !!section.plugin_identifier;
    return isCollapsible && !isSpecialSectionType(section.type) && !hasPlugin;
  }

  isCommentable() {
    const {section} = this.props;
    const hasPlugin = !!section.plugin_identifier;
    return (this.props.canComment && !hasPlugin);
  }

  render() {
    const {section, user} = this.props;
    const {language} = this.context;
    const collapsible = this.isCollapsible();
    const collapsed = collapsible && this.state.collapsed;
    const titleDiv = this.getTitleDiv(collapsed, collapsible);
    let commentList = null;
    if (collapsed) {
      return (
        <div className="section-list-item">
          <div className="section-list-item-image" onClick={this.toggle.bind(this)}>
            {section.images.length ? <img alt="" src={section.images[0].url} /> : null}
          </div>
          <div className="section-list-item-content">
            {titleDiv}
            <div className="section-abstract" dangerouslySetInnerHTML={{__html: getAttr(section.abstract, language)}} />
          </div>
          <hr/>
        </div>
      );
    }

    if (!isSpecialSectionType(section.type)) {
      commentList = (<SortableCommentList
        section={section}
        comments={(this.props.comments ? this.props.comments.results : null) || []}
        canComment={this.isCommentable() && userCanComment(this.props.user, section)}
        onPostComment={this.onPostComment.bind(this)}
        canVote={this.props.canVote}
        onPostVote={this.onPostVote.bind(this)}
        canSetNickname={!user}
        isSectionComments={section}
        onDeleteComment={this.props.handleDeleteClick}
        onEditComment={this.props.onEditComment}
      />);
    }
    const imageList = getImageList(section, language);
    const sectionClass = classNames({
      'hearing-section': true,
      'closure-info': section.type === "closure-info"
    });
    const pluginContent = this.renderPluginContent(section);
    return (<div className={sectionClass}>
      {titleDiv}
      <div className="section-content">
        {imageList}
        {section.type !== "main" ?
          <div
            className="section-abstract lead"
            dangerouslySetInnerHTML={{__html: getAttr(section.abstract, language)}}
          /> :
          null}
        <div dangerouslySetInnerHTML={{__html: getAttr(section.content, language)}} />
        {pluginContent}
      </div>
      {commentList}
      <hr/>
    </div>);
  }
}

Section.defaultProps = {
  showPlugin: true,
  isCollapsible: true,
};

Section.propTypes = {
  canComment: React.PropTypes.bool,
  canVote: React.PropTypes.bool,
  comments: React.PropTypes.object,
  isCollapsible: React.PropTypes.bool,
  linkTo: React.PropTypes.oneOfType([
    React.PropTypes.string,
    React.PropTypes.object,
    React.PropTypes.func
  ]),
  loadSectionComments: React.PropTypes.func,
  onPostComment: React.PropTypes.func,
  onPostVote: React.PropTypes.func,
  section: React.PropTypes.object.isRequired,
  showPlugin: React.PropTypes.bool,
  user: React.PropTypes.object,
  handleDeleteClick: React.PropTypes.func,
  onEditComment: React.PropTypes.func,
};

Section.contextTypes = {
  language: React.PropTypes.string.isRequired
};
