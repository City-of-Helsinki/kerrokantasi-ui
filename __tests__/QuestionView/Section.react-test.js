import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {Section} from '../../src/components/Section';
import {MemoryRouter} from 'react-router-dom';
import {mockStore, getIntlAsProp} from '../../test-utils';

// Renders the Hearings component using enzymes shallow rendering.
// You can pass props you want to override as a parameter.
const setup = propOverrides => {
  const {sectionComments, hearingLists: {allHearings}, hearingWithSections, dispatch} = mockStore;
  const props = Object.assign({
    match: {
      params: {
        hearingSlug: allHearings.data[0].slug
      }
    },
    location: {
      pathname: '/' + allHearings.data[0].slug
    },
    hearingSlug: allHearings.data[0].slug,
    sectionId: hearingWithSections.mockHearingWithSections.data.sections[0].id,
    comments: sectionComments,
    dispatch,
    isCollapsible: false,
    canVote: true,
    canComment: true,
    section: hearingWithSections.mockHearingWithSections.data.sections[0],
    language: 'fi',
    sectionNav: {
      shouldShowBrowser: true
    }
  }, propOverrides);

  const wrapper = shallow(<MemoryRouter><Section intl={getIntlAsProp()} {...props} /></MemoryRouter>);

  return {
    props,
    wrapper
  };
};

test('Section component should render as expected', () => {
  const {wrapper} = setup();
  const tree = toJson(wrapper.dive().dive());
  expect(tree).toMatchSnapshot();
});
