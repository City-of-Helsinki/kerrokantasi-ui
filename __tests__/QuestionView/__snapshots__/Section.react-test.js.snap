// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section component should render as expected 1`] = `
<div
  className="hearing-section"
>
  <div
    className="section-content"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Samanaikaisesti Isosaaren avaamisen kanssa Helsingin kaupunki aloittaa saaren asemakaavoituksen. Saarta suunnitellaan monipuoliseksi virkistyksen ja matkailun keitaaksi. &nbsp;&nbsp;&nbsp;</p>
<p>Kerro ajatuksesi Isosaaren kehittämisestä: &nbsp;</p>
<ul>
  <li>Minkälaisia matkailupalveluja ja toimintaa Isosaaressa voisi olla?</li>
  <li>Mikä Isosaaressa on ainutlaatuista?&nbsp;</li>
  <li>Millainen Isosaari olisi tulevaisuudessa? Ideoita ja visioita käytetään saaren suunnittelun tukena ja innoituksena.&nbsp;</li>
</ul>
<p>&nbsp;&nbsp;Ideoita ja visioita käytetään saaren suunnittelun tukena ja innoituksena.</p>
<p>Lisätietoja:</p>
<p><a href=\\"http://www.visitisosaari.fi\\">Tietoa Isosaaressa vieraileville</a></p>",
        }
      }
    />
  </div>
  <Connect(InjectIntl(SortableCommentList))
    canComment={true}
    canVote={true}
    isSectionComments={
      Object {
        "abstract": Object {
          "fi": "Tämä on suomenkielinen testiabstrakti. Sisältää huikean kuvauksen kyseisestä kuulemisesta.",
        },
        "commenting": "open",
        "content": Object {
          "fi": "<p>Samanaikaisesti Isosaaren avaamisen kanssa Helsingin kaupunki aloittaa saaren asemakaavoituksen. Saarta suunnitellaan monipuoliseksi virkistyksen ja matkailun keitaaksi. &nbsp;&nbsp;&nbsp;</p>
<p>Kerro ajatuksesi Isosaaren kehittämisestä: &nbsp;</p>
<ul>
  <li>Minkälaisia matkailupalveluja ja toimintaa Isosaaressa voisi olla?</li>
  <li>Mikä Isosaaressa on ainutlaatuista?&nbsp;</li>
  <li>Millainen Isosaari olisi tulevaisuudessa? Ideoita ja visioita käytetään saaren suunnittelun tukena ja innoituksena.&nbsp;</li>
</ul>
<p>&nbsp;&nbsp;Ideoita ja visioita käytetään saaren suunnittelun tukena ja innoituksena.</p>
<p>Lisätietoja:</p>
<p><a href=\\"http://www.visitisosaari.fi\\">Tietoa Isosaaressa vieraileville</a></p>",
        },
        "created_at": "2017-10-23T11:07:26.668934Z",
        "id": "7CxRxrlvrW9uCCjEaB9rW1dENHQMudeK",
        "images": Array [
          Object {
            "caption": Object {
              "fi": "A Wonderfull caption",
            },
            "height": 631,
            "id": 92,
            "title": Object {},
            "url": "https://api.hel.fi/kerrokantasi-test/media/images/2017/10/LzqVMrMI.jpeg",
            "width": 893,
          },
        ],
        "n_comments": 1,
        "plugin_data": "",
        "plugin_fullscreen": false,
        "plugin_identifier": "",
        "plugin_iframe_url": "",
        "published": false,
        "title": Object {},
        "type": "main",
        "type_name_plural": "pääosiot",
        "type_name_singular": "pääosio",
        "voting": "registered",
      }
    }
    sectionNav={
      Object {
        "shouldShowBrowser": true,
      }
    }
  />
</div>
`;
